### PR TITLE
Support partial texture updates via efb copies

### PR DESCRIPTION
**Pull request by mimimi085181**
New Super Mario Bros uses a 1024x1024 texture for most objects in the levels. This texture is generated from data in memory, but some parts of it are generated via efb copies. On efb to texure only, the data from the efb copies is not written to memory. This results in those parts being either static or missing in Dolphin's representation of the texture. Examples for this are the spinning coins, the blue coins and the question mark block animation.

This pr copies the graphics from these efb copies into the graphics of of the big texture, at the right position. After this, the data in memory is still wrong, but the textures on the GPU are correct. Since this is only a display issue, that's fine.

Note: Apparently the efb copies happen 1st, and then the big texture is loaded, which is why the updates are applied on load, and not on the efb copy.

The limitations of this pr are(not going to be fixed in this pr):
- The big texture is not upscaled, so the partial texture updates are disabled, if the IR is not 1 and scaled efb copies is enabled. Otherwise there would be big graphical issues.
- The updates are disabled for efb to ram. 1. Without upscaling it would identical to how it looks now and 2. this would require a change or the removal of MakeRangeDynamic, among other things.
- Paletted textures are excluded, because support for those would require more work, like converting the efb copies before using them for the updates. This needs an example where this would be useful.
- Efb copies are excluded. Would this even make sense somehow?

About 90% of the OpenGL part was written by HdkR, thanks a lot.

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Can't really say much about the code itself; but style-wise I'd recommend renaming the methods to conform to the same style as the others.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by Linktothepast** [comment]
Virtual console nes games seem to have garbled graphics with d3d in this build.

**Comment by mimimi085181** [comment]
Can you please test if this happens on master? I got the same bug with 4.0-6759, but not with 4.0-6749.

**Comment by JMC47** [comment]
NES games have never worked on D3D for me on NVIDIA.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 1]
```diff
@@ -238,6 +294,8 @@ static u32 CalculateLevelSize(u32 level_0_size, u32 level)
 // Used by TextureCache::Load
 TextureCache::TCacheEntryBase* TextureCache::ReturnEntry(unsigned int stage, TCacheEntryBase* entry)
 {
+	entry->do_partial_texture_updates();
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Why in ReturnEntry? IMO this code should be called in CopyRenderTargetToTexture

**Comment by Linktothepast** [comment]
Yes seems a recent regression in master. What was the merge stable branch about?

**Comment by JMC47** [comment]
I'mma take a blind guess and say it probably shouldn't have happened.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -211,6 +211,62 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+
+	u32 numBlocksX = expanded_width / block_width;
+	//u32 numBlocksY = expanded_height / block_height;
+
+	u32 cacheLinesPerRow;
+	// TODO: Is this check correct?
+	if (format == GX_TF_RGBA8) // RGBA takes two cache lines per block; all others take one
+		cacheLinesPerRow = numBlocksX * 2;
+	else
+		cacheLinesPerRow = numBlocksX;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think this should be replaced by SizeInBytesPerBlock, maybe a new function within https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/TextureDecoder.h#L64

So mystride = SizeInBytesPerBlock * numBlocksX;

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 3]
```diff
@@ -211,6 +211,62 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+
+	TexCache::iterator iter = textures_by_address.lower_bound(addr);
+	TexCache::iterator iterend = textures_by_address.upper_bound(addr + size_in_bytes);
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This doesn't find textures which starts below addr. As we're only able to search by the base address of the textures, we likely have to start at addr - max_texture_size.
IIRC 1024*1024*4 is the max texture size.

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Can we do this in another pr? For now i only want to do this for efb copies that are completely within the other texture.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 4]
```diff
@@ -211,6 +211,62 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+			&& addr <= entry->addr && entry->addr + entry->size_in_bytes <= addr + size_in_bytes
+			&& !(native_width == entry->native_width && native_height == entry->native_height))
+		{
+			//TODO: Is this calculation correct? Where does the 8 come from?
+			u32 data_offset = entry->addr - addr;
+
+			u32 block_y = data_offset / (cacheLinesPerRow * 32);
+			u32 y = block_y * block_height;
+
+			u32 block_x = (data_offset % (cacheLinesPerRow * 32)) / numBlocksX * 8;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: There is no need to divide by numBlocksX at all.

**Comment by Linktothepast** [comment]
@JMC47 It seems it is about anisotropic filtering and aniso 1x acts like a higher number (2x etc.) in a previous build. Sorry for the offtopic.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 1 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This depends on the order in which things are happening, which i don't know. The big texture in new super mario bros gets changed for different worlds(or maybe even single stages). When does this happen, and when do the texture updates via efb copies happen?

Or more precise, what's there 1st? The efb copies or the texture?

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 4 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: (data_offset % (cacheLinesPerRow * 32) is 7680 in one example, and i want to get to 960 from there. The block_y and y calculation kinda makes sense to me, because the efb2ram code uses cacheLinesPerRow * 32 in a similar way. I'm not sure about the block_x and x calculation, for me it makes sense. but just not the 8.

**Comment by mimimi085181** [comment]
I think it's about the centroid pr, but enough off topic please.

**Comment by mimimi085181** [comment]
@BhaaLseN: What names do you suggest?

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 4 above)
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Of course, you need to divide by the block size

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 2 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think this is something for a different pr as the calculation is used for efb2ram as well. And then we got the problem with the different formats...

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 3 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: BTW: I think max texture size is 1024 * 1024 * max(texelsizeinnibbles)/2 so 4194304. This would be tiny speedup for MakeRangeDynamic...

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 1 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Ok, i wrote a small test program for this, and this is the debug output:
[code]
11:05:590 D3DBase.cpp:113 N[Video]: Successfully loaded D3DCompiler_42.dll. If you're having trouble, try updating your DX runtime first.
11:05:904 BPStructs.cpp:676 W[Video]: Unknown BP opcode: address = 0x00000069 value = 0x000004ed
11:05:904 BPStructs.cpp:676 W[Video]: Unknown BP opcode: address = 0x00000046 value = 0x00000273
11:49:874 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10334860
11:49:874 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10344760
11:49:875 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10334760
11:49:875 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10343860
11:49:875 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10343960
11:49:875 TextureCache.cpp:183 E[Video]: efb copy offset: 0x10343a60
11:49:889 TextureCacheBase.cpp:418 E[Video]: normal texture offset: 0x10332960
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10334760, w/h: 1024/1024, update at: x/y:  960/   0, w/h:   32/  32
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10334860, w/h: 1024/1024, update at: x/y:  992/   0, w/h:   32/  32
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10343860, w/h: 1024/1024, update at: x/y:  480/  32, w/h:   32/  32
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10343960, w/h: 1024/1024, update at: x/y:  512/  32, w/h:   32/  32
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10343a60, w/h: 1024/1024, update at: x/y:  544/  32, w/h:   32/  32
11:49:895 TextureCacheBase.cpp:259 E[Video]: Doing partial texture update for a texture at: 0x10332960, efb copy at: 0x10344760, w/h: 1024/1024, update at: x/y:  960/  32, w/h:   32/  32
[/code]

So the efb copies are apparently done 1st, and then we load the texture in:
TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)

So the way the pr currently is implemented seems fine to me. The textrue updates are uploaded by the game, then we load the texture, maybe rebuild it for a new world/stage, and then we apply the updates on top of it.

But ReturnEntry might really be a bad place for this. Should i move it to the places before calling ReturnEntry where it makes sense to try this?

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 4 above)
Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Thanks, i think i fixed this now:
u32 block_size = block_width * block_height * TexDecoder_GetTexelSizeInNibbles(format) / 2;
u32 block_x = (data_offset % (cacheLinesPerRow * 32)) / block_size;
u32 x = block_x * block_width;

**Comment by BhaaLseN** [comment]
At least ```DoPartialTextureUpdate``` instead of ```do_partial_texture_update``` (with the underscores); no idea if there would be a more useful name for those otherwise.

**Comment by mimimi085181** [comment]
So i tried to address all comments. Next round of reviews please.

Also, thanks to HdkR, the pr now supports OpenGL as well.

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 2 above)
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I don't think so, as you've implemented it already: u32 block_size = block_width * block_height * TexDecoder_GetTexelSizeInNibbles(format) / 2;

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 5]
```diff
@@ -211,6 +211,62 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+		if (entry->IsEfbCopy() && entry->frameCount == FRAMECOUNT_INVALID && entry->copyMipMapStrideChannels == cacheLinesPerRow
+			&& addr <= entry->addr && entry->addr + entry->size_in_bytes <= addr + size_in_bytes
+			&& !(native_width == entry->native_width && native_height == entry->native_height))
+		{
+			u32 data_offset = entry->addr - addr;
+
+			u32 block_y = data_offset / (cacheLinesPerRow * 32);
+			u32 y = block_y * block_height;
+
+			u32 block_size = block_width * block_height * TexDecoder_GetTexelSizeInNibbles(format) / 2;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Use this block size more often, eg by:
´´´
u32 block_offset = (entry->addr - addr) / block_size;
u32 block_x = block_offset % numBlocksX;
u32 block_y = block_offset / numBlocksX;
´´´

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
I would request by default we disable scaled EFB Copies in New Super Mario Bros. Wii in the INI until a scaling version is created.

Are there any other reasons not to merge this?  Tested pretty thoroughly, works great.

**Comment by mimimi085181** [comment]
I need to edit at least some comments before that. And there are likely to be more coding style issues.

**Comment by JMC47** [comment]
Okay, just wanted to hear. @degasus get reviewin'!

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 6]
```diff
@@ -211,6 +211,54 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+		|| (g_ActiveConfig.bCopyEFBScaled && g_ActiveConfig.iEFBScale != SCALE_1X))
+		return;
+
+	u32 block_width = TexDecoder_GetBlockWidthInTexels(format);
+	u32 block_height = TexDecoder_GetBlockHeightInTexels(format);
+	u32 block_size = block_width * block_height * TexDecoder_GetTexelSizeInNibbles(format) / 2;
+
+	u32 expanded_width = (native_width + block_width - 1) & ~(block_width - 1);
+
+	u32 numBlocksX = expanded_width / block_width;
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: u32 numBlocksX = (native_width + block_width - 1) / block_width;

There is no need to also calculate expanded_width.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 7]
```diff
@@ -211,6 +211,54 @@ bool TextureCache::TCacheEntryBase::OverlapsMemoryRange(u32 range_address, u32 r
...
+			&& !(native_width == entry->native_width && native_height == entry->native_height))
+		{
+			u32 block_offset = (entry->addr - addr) / block_size;
+			u32 block_x = block_offset % numBlocksX;
+			u32 block_y = block_offset / numBlocksX;
+
+			u32 x = block_x * block_width;
+			u32 y = block_y * block_height;
+
+			ERROR_LOG(VIDEO, "Doing partial texture update for a texture at: 0x%08x, efb copy at: 0x%08x, w/h: %4d/%4d, update at: x/y: %4d/%4d, w/h: %4d/%4d", addr, entry->addr, native_width, native_height, x, y, entry->native_width, entry->native_height);
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This is no error.

Comment (mimimi085181) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: It will be removed before merging. With logging enabled, the 1000s of messages kill the performane of the emulator. Like after playing 1-2 levels in New Super Mario Bros, the fps went down to 40-50 for me, with framelimt auto. While holding hold tab in that situation, it only got to 200 fps, so i think there's some issue with the log. It was even only set to log to window, and the window was not even active.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by degasus** [comment]
The code looks goot, but I still dislike the fact that this way will spam lots of never usable textures within the texture cache.

efb copys which have a "broken" stride should not be copied into the cache, they should just be applied to other textures. But I agree, @mimimi085181 has a point about the order. Maybe fall back to efb2ram if there is no texture to update? For NSMBW, this won't matter (but maybe for the first frame...)

**Comment by mimimi085181** [comment]
You would really like to apply these updates at the time of the efb copy, do you? I see many problems with that: What would you do in this case:
texture is loaded -> efb copy with updates -> texture update(for a new stage for example) and load
Most likely Dolphin would end up discarding the updates from the efb copies, every time the texture itself is updated. And if a game constantly updates the texture via normal means and efb copies at the same time, it wouldn't work at all.

Falling back to efb to ram, when there is no texture to update, that sounds like a complicated solution. And it wouldn't work well with upscaling. (assuming that upscaling efb copies works better than the yet to be implemented normal texture upscaling)

To skip creating the efb copy as standalone texture, before copying it into the normal texture, that would be nice for performance and memory usage, but i don't see how it should be done. As far as i can see, that's the only advantage it would have. For me, the memory usage for these small texture is not a concern. And the performance isn't either, because this is still 3 times as fast as efb 2 ram.

Are you talking about the efb copy spam problem, that unused efb copies stay forever in the cache? I think this should be addressed in its own pr, it's a big problem that is not limited to this special kind of efb copies. Well, we could delete them after applying, instead of just marking them as used. But i don't know if a game would do anything unusual like using such an efb copy as standalone texture as well.

**Comment by degasus** [comment]
Yes, I really like to :D

"And if a game constantly updates the texture via normal means and efb copies at the same time, it wouldn't work at all." -- This is likely by design. How do you check which region got overwritten? Maybe it also affected the region of our old efb2tex call. The current code may just reuse any old (partial) efb copy for a completely overwritten texture. IMO it's more important to not overwrite with the wrong values than to loose upscaling support. And for NSMBW, upscaling will work as well starting with the second efb copy ;)

"i don't know if a game would do anything unusual like using such an efb copy as standalone texture as well" If the stride doesn't match, this texture can't be used directly.

**Comment by JMC47** [comment]
I'm guessing this is hung up implementation wise then?

**Comment by degasus** [comment]
I think it's fine for now. If I want it in the opposite way, I can still swap it again :D

**Comment by Linktothepast** [comment]
Should this be enabled by default in the gameini of nsmb? I mean disabling scaled efb copies so that it will be enabled with efb to texture.

**Comment by JMC47** [comment]
I lean toward yes, but, I don't know everyone else's opinions.
